src/persist/src/sqlite.rs

```diff
@@ -62,32 +61,42 @@ impl FromSql for SeqNo {
 /// Implementation of [Consensus] over a sqlite database.
 #[derive(Debug)]
 pub struct SqliteConsensus {
-    // N.B. tokio::sync::mutex seems to cause deadlocks.  See #12231.
+    // Intentionally a std::sync::Mutex instead of a tokio one because we don't
+    // hold it across await points (ha! there are no await points with
+    // rusqlite).
     conn: Arc<Mutex<Connection>>,
```

Comment:
the structure of this code now with the spawn_blocking calls makes it pretty obvious that this mutex linearizes every call into SqliteConsensus. specifically, there's no key-level granularity, which might be (one of) our problem(s) we're seeing in 12216's CI with 22 tables each trying to advance the upper once per second. a fix for that feels like it should be a separate PR/commit though